File: LEETCODE/98.cpp
//98. Validate Binary Search Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        
        vector<int> res;
        checkBST(root,res);
        
        for(int i=1;i<res.size();i++){
            if(res[i]<=res[i-1])
                return false;
        }
        
        return true;
        
    }
    
    void checkBST(TreeNode * root,vector<int> &res){
        if(!root)
            return;
        
        checkBST(root->left,res);
        res.push_back(root->val);
        checkBST(root->right,res);
    }
};

File: LEETCODE/LEETCODE-PYTHON/98.py
# 98. Validate Binary Search Tree
# https://leetcode.com/problems/validate-binary-search-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Method-1 using recursion (TC-O(n))
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def helper(node, low=-math.inf, high=math.inf):
            if node is None:
                return True

            if node.val <= low or node.val >= high:
                return False

            return helper(node.right, node.val, high) and helper(node.left, low, node.val)

        return helper(root)
            
